_overviews/tutorials/binary-compatibility-for-library-authors.md

```diff
@@ -0,0 +1,221 @@
+The library's bytecode is referenced by its class/method signatures and loaded lazily
+by the JVM classloader during runtime. If a class or method matching the signature is not found,
+an exception is thrown. 
+
+As a result of this execution model:
+
+* We need to provide the JARs of every library used in our dependency tree when starting an application since the library's bytecode is only referenced, not merged into its user's bytecode
+* A missing class/method problem may only surface after the application has been running for a while, due to lazy loading.
+
+Common exceptions from classloading failures includes 
+`InvocationTargetException`, `ClassNotFoundException`, `MethodNotFoundException`, and `AbstractMethodError`.
+
+Let's illustrate this with an example:
+
+Consider an application `App` that depends on `A` which itself depends on library `C`. When starting the application we need to provide the class files
+for all of `App`, `A` and `C` (something like `java -cp App.jar:A.jar:C.jar:. MainClass`). If we did not provide `C.jar` or if we provided a `C.jar` that does not contain some classes/methods
+which `A` calls, we will get classloading exceptions when our code attempts to invoke the missing classes/methods.
+
+These are what we call **Linkage Errors** -- errors that happen when the compiled bytecode references a name that cannot be resolved during runtime.
+
+### What about Scala.js and Scala Native?
+Similarly to the JVM, Scala.js and Scala Native have their respective equivalents of .class files, namely .sjsir files and .nir files. Similarly to .class files, they are distributed in .jars and linked together at the end.
+
+However, contrary to the JVM, Scala.js and Scala Native link their respective IR files at link time, so eagerly, instead of lazily at run-time. Failure to correctly link the entire program results in linking errors reported while trying to invoke `fastOptJS`/`fullOptJS` or `nativeLink`.
+
+Besides that difference in the timing of linkage errors, the models are extremely similar. **Unless otherwise noted, the contents of this guide apply equally to the JVM, Scala.js and Scala Native.**
+
+Before we look at how to avoid binary incompatibility errors, let us first
+establish some key terminologies we will be using for the rest of the guide.
+
+## What are Evictions, Source Compatibility and Binary Compatibility?
+
+### Evictions
+When a class is needed during execution, the JVM classloader loads the first matching class file from the classpath (any other matching class files are ignored).
+Because of this, having multiple versions of the same library in the classpath is generally undesirable:
+
+* Need to fetch and bundle multiple library versions when only one is actually used
+* Unexpected runtime behavior if the order of class files changes
+
+Therefore, when resolving JARs to use for compilation and packaging, most build tools will pick only one version of each library and **evict** the rest.
```

Comment:
It might be useful to explain that the default scheme for this in sbt (and gradle, and many other build tools) is to take the latest version, since most libraries are backwards but not forwards compatible.

Reply:
I did initially mention this but was asked to remove it. What do you think @jvican?

Reply:
I believe you can add "The most common solution in build tools is to take the latest version (sbt, gradle, etc)." and will still be good 👍

Reply:
done